cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/SystemArguments.java

```diff
@@ -236,6 +239,32 @@ public static Resources getResources(Map<String, String> args, @Nullable Resourc
   }
 
   /**
+   * Extracts and returns a configuration map containing container related Twill settings.
+   *
+   * @param args the arguments to use for looking up resources configurations
+   * @param containerMemory the container memory size in MB
+   * @return a map of configurations
+   */
+  public static Map<String, String> getTwillContainerConfigs(Map<String, String> args, int containerMemory) {
+    Integer reservedMemory = getPositiveInt(args, RESERVED_MEMORY_KEY, "reserved memory size");
+    if (reservedMemory == null) {
+      return Collections.emptyMap();
+    }
+    int heapMemory = containerMemory - reservedMemory;
+    if (heapMemory <= 0) {
+      LOG.warn("Ignoring invalid reserved memory size '{}' from runtime arguments. " +
+                 "It must be smaller than container memory size '{}'", reservedMemory, containerMemory);
+      return Collections.emptyMap();
+    }
+    double ratio = ((double) heapMemory) / containerMemory;
+
+    Map<String, String> config = new HashMap<>();
+    config.put(Configs.Keys.JAVA_RESERVED_MEMORY_MB, reservedMemory.toString());
+    config.put(Configs.Keys.HEAP_RESERVED_MIN_RATIO, String.format("%.2f", ratio));
```

Comment:
So you allow configuring the reserved memory and then you derive the heap ratio from that? Another approach would be to allow the user to specify both. Then it would behave the same way as in the CConf. I would find that more intuitive and easier to document

Reply:
I have been thinking about the use case for this (running a program with specific memory setting for different container types of that program).

We have the `ratio` setting in cConf is mainly because the setting applies to all containers, regardless of the container memory size. 

However, in program runtime case, the user already controlled both the container memory size and the reserved memory size, hence having the `ratio` settable is actually redundant and is harder for the user to understand what's going on.

In terms of documentation, I think by just saying setting `system.resources.memory` for container size and `system.resources.reserved.memory` for reserved (non-heap) memory is easier to explain (-Xmx heap size = container size - reserved size), comparing to having `ratio` in place, which is determined by:

```
  public static int computeMaxHeapSize(int containerMemory, int nonHeapMemory, double minHeapRatio) {
    if (((double) (containerMemory - nonHeapMemory) / containerMemory) >= minHeapRatio) {
      // Reduce -Xmx by the reserved memory size.
      return containerMemory - nonHeapMemory;
    } else {
      // If it is a small VM, just discount it by the min ratio.
      return (int) Math.ceil(containerMemory * minHeapRatio);
    }
  }
```

Reply:
Except if the user sets a preference for a namespace. Then that applies to all containers in that namespace, and it should behave exactly the same as in the CConf, right?

Reply:
I think allowing setting `ratio` or have it derived have their own pros and cons. I think basically having `ratio` allows easier setting at higher level, regardless of what the individual container size. In contrast, honoring reserved memory setting directly makes it easier to override at specific task level and easier to explain.

Let's say we allow setting `ratio`.

At namespace level:
`system.resources.memory = 1024m`
`system.resources.reserved.memory = 512M`
`system.resources.heap.memory.ratio = 0.7`

At program level:
`system.resources.memory = 512M`

Then the program would automatically have `-Xmx=358M`. However, if the one want specify the max heap size, then he has to override two settings (`ratio` and `reserved`), instead of one, and also need to calculate a `ratio` value (or a small enough value).

Also, maybe we should have the setting more direct as `system.resources.heap.memory` which maps to `-Xmx` directly?